src/server/async.rs

```diff
@@ -0,0 +1,109 @@
+//! [server::async](/src/server/async.rs)
+//!
+//! This module contains the procedures for spinning up a lightweight mock websocket server
+//! using tokio and tunngstenite.
+#[allow(unused_import_braces, unused_imports)]
+#[rustfmt::skip]
+use {
+    crate::errors::external::WebmocketErr,
+    futures_util::{SinkExt, StreamExt},
+    pin_project_lite::pin_project,
+    std::borrow::Borrow,
+    std::sync::atomic::{AtomicU16, Ordering::{Acquire, Relaxed}},
+    std::pin::Pin,
+    tokio::net::{TcpListener, TcpStream, ToSocketAddrs},
+    tokio::task::JoinHandle,
+    tokio_tungstenite::accept_hdr_async,
+    tokio_tungstenite::tungstenite::{
+        connect,
+        handshake::server::{Request, Response},
+        Message,
+    },
+};
+
+// A non-owning handle to a mock websocket server managed by tokio and tunngstenite.
+pin_project! {
+    pub struct Webmocket {
+        port: &'static AtomicU16,
+        #[pin]
+        pub handle: Pin<Box<JoinHandle<()>>>,
```

Comment:
It seems the heap `Pin` isn't necessary since the spawned tasks live in the internal task queues. Further, the future has to be `Send` since work-stealing may move them to another thread and already-spawned futures need to be valid for `'static` b/c they free-run after drop until the process exits.